### PR TITLE
Hate crime data download

### PR DESCRIPTION
**Pull request by rbhoot**
The hate crime data is available at https://ucr.fbi.gov/hate-crime . The data is divided into 14 different tables(and a 15th participation details table) for each year. The scripts goes through the range of input years and collects the download links for all the tables, finally downloading and storing them.
NOTE: the requirements.txt is modified to add [beautiful soup](https://www.crummy.com/software/BeautifulSoup/) package to parse through the pages.

### CODE REVIEW

**Review by Abilityguy — APPROVED** [review]
LGTM

File: scripts/fbi/hate_crime/download_publication_data.py [hunk 1]
```diff
@@ -0,0 +1,363 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""A script to scrape and download FBI Hate Crime table publications data."""
+import requests
+import os
+from bs4 import BeautifulSoup
```
Comment (ajaits) [inline scripts/fbi/hate_crime/download_publication_data.py:17]: please group imports into the following blocks, each sorted alphabetically if there are no interdependancies:


import abc
import def

from ghi import myfunc
from lmn import pqr

File: scripts/fbi/hate_crime/download_publication_data.py [hunk 2]
```diff
@@ -0,0 +1,363 @@
...
+flags.DEFINE_integer('end_year', 2019, 'End year(Min. 2005)', lower_bound=2005)
+flags.DEFINE_boolean('force_fetch', True,
+                     'forces api query and not return cached result')
+flags.DEFINE_string('store_path', './tmp', 'Path to store the output')
+flags.DEFINE_string('cache_path', './tmp/.cache_html/',
+                    'Path to store the cached html files')
+flags.DEFINE_string('base_url', 'https://ucr.fbi.gov/hate-crime',
+                    'The base url containing list of URLs for each year')
+
+# TODO(issue id here) store http errors
```
Comment (ajaits) [inline scripts/fbi/hate_crime/download_publication_data.py:38]: Pls open an issue for these.

File: scripts/fbi/hate_crime/download_publication_data.py [hunk 3]
```diff
@@ -0,0 +1,363 @@
...
+# Headers to make request
+_HEADERS = {
+    'User-Agent':
+        'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/96.0.4664.55 Safari/537.36'
+}
+"""
+The file names of excel files changes across the years.
+The URLs to access table appear in sequence of table number,
+hence code uses enumerate to assign the table number.
+years_with_no_table0 list help make this correction.
```
Comment (ajaits) [inline scripts/fbi/hate_crime/download_publication_data.py:50]: can reword `make this correction` to `skip table-0 for specific years where it is not available`

File: scripts/fbi/hate_crime/download_publication_data.py [hunk 4]
```diff
@@ -0,0 +1,363 @@
...
+  """
+    req = cur_session.get(url)
+    if req.status_code == requests.codes.ok:
+        resp_html = req.text
+        resp_soup = BeautifulSoup(resp_html, features="html.parser")
+    else:
+        resp_html = None
+        resp_soup = None
+
+        print('HTTP status code: ' + str(req.status_code))
```
Comment (ajaits) [inline scripts/fbi/hate_crime/download_publication_data.py:101]: Can remove empty line 100 and add an ERROR prefix to the print mesage with the url attempted.

File: scripts/fbi/hate_crime/download_publication_data.py [hunk 5]
```diff
@@ -0,0 +1,363 @@
...
+        year_str: string version of the year for which the URL is to be found.
+    Returns:
+        URL to the page for the data of particular year.
+  """
+    year_url = None
+    base_a_list = html_soup.find(id='content-core').find_all('a')
+    for cur_a in base_a_list:
+        cur_year = cur_a.string
+        if cur_year == year_str:
+            print(cur_year)
```
Comment (ajaits) [inline scripts/fbi/hate_crime/download_publication_data.py:118]: Pls remove print or add a more descriptive message.

File: scripts/fbi/hate_crime/download_publication_data.py [hunk 6]
```diff
@@ -0,0 +1,363 @@
...
+  
+  The method uses status variables set in the status_dict for context.
+    Returns:
+        URL to the page with list of all the tables for a particular year.
+  """
+    global _STATUS_DICT
+    # fetch or return from cache
+    if _STATUS_DICT['force_fetch'] or not os.path.isfile(
+            os.path.join(_STATUS_DICT['cache_dir'],
+                         f"{_STATUS_DICT['cur_year']}.html")):
```
Comment (ajaits) [inline scripts/fbi/hate_crime/download_publication_data.py:134]: `os.path.join(_STATUS_DICT['cache_dir'],  f"{_STATUS_DICT['cur_year']}.html")` is used multiple times.
Can set into a local var instead.

File: scripts/fbi/hate_crime/download_publication_data.py [hunk 7]
```diff
@@ -0,0 +1,363 @@
...
+                             f"{_STATUS_DICT['cur_year']}.html"), 'r') as fp:
+            year_html = fp.read()
+            year_soup = BeautifulSoup(year_html, features="html.parser")
+
+    # some year pages contain and iframe which has the actual html content,
+    # 1st iframe is used if multiple are found
+    year_iframe_list = year_soup.find_all('iframe')
+    if len(year_iframe_list) > 0:
+        _STATUS_DICT['cur_year_url'] = year_iframe_list[0]['src']
+        print('iframe', _STATUS_DICT['cur_year_url'])
```
Comment (ajaits) [inline scripts/fbi/hate_crime/download_publication_data.py:153]: Pls remove print or add more context, `f'downloading iframe: <url>'`

File: scripts/fbi/hate_crime/download_publication_data.py [hunk 8]
```diff
@@ -0,0 +1,363 @@
...
+                os.path.join(_STATUS_DICT['cache_dir'],
+                             f"{_STATUS_DICT['cur_year']}.html"), 'w') as fp:
+            fp.write(year_html)
+    # convert for use of base URL in the next step
+    _STATUS_DICT['cur_year_url'] = _STATUS_DICT[
+        'cur_year_url'][:_STATUS_DICT['cur_year_url'].rfind('/')]
+    year_subitem_list = year_soup.find_all('div', class_='subitem')
+    access_tables_found = False
+    access_url = None
+    if len(year_subitem_list) > 0:
```
Comment (ajaits) [inline scripts/fbi/hate_crime/download_publication_data.py:166]: can remove 160 so save an indent level as the for loop will automatically skip if list is empty

File: scripts/fbi/hate_crime/download_publication_data.py [hunk 9]
```diff
@@ -0,0 +1,363 @@
...
+        for year_subitem in year_subitem_list:
+            if not access_tables_found:
+                all_a = year_subitem.find_all('a')
+                for cur2_a in all_a:
+                    access_url = cur2_a.get('href')
+                    if not access_tables_found:
```
Comment (ajaits) [inline scripts/fbi/hate_crime/download_publication_data.py:172]: Can save an indent level with 
```if access_tables_found:
  break
```

File: scripts/fbi/hate_crime/download_publication_data.py [hunk 10]
```diff
@@ -0,0 +1,363 @@
...
+
+    tables_div_all = access_soup.find(id='col_b')
+    if not tables_div_all:
+        tables_div_all = access_soup.find(id='col_2')
+    tables_a_all = tables_div_all.find_all('a')
+    if _STATUS_DICT['cur_year'] in _YEARS_WITH_NO_TABLE0:
+        _STATUS_DICT['table_html_url_list'] = ['']
+    else:
+        _STATUS_DICT['table_html_url_list'] = []
+    for i, cur_table_a in enumerate(tables_a_all):
```
Comment (ajaits) [inline scripts/fbi/hate_crime/download_publication_data.py:228]: `i` seems unused in the loop. Can avoid enumerate?

File: scripts/fbi/hate_crime/download_publication_data.py [hunk 11]
```diff
@@ -0,0 +1,363 @@
...
+    return _STATUS_DICT['table_html_url_list']
+
+
+def _find_table_xls_url():
+    """Find the URL list of XLS version of all the tables and download them.
+  """
+    global _STATUS_DICT
+    for table_i, table_html_url in enumerate(
+            _STATUS_DICT['table_html_url_list']):
+        if table_html_url:
```
Comment (ajaits) [inline scripts/fbi/hate_crime/download_publication_data.py:242]: save an indent with: 
```
if not table_html_url:
   continue
```

File: scripts/fbi/hate_crime/download_publication_data.py [hunk 12]
```diff
@@ -0,0 +1,363 @@
...
+            print(table_html_url)
```
Comment (ajaits) [inline scripts/fbi/hate_crime/download_publication_data.py:243]: Please add more context: say, 'Downloading table from {url}'

File: scripts/fbi/hate_crime/download_publication_data.py [hunk 13]
```diff
@@ -0,0 +1,363 @@
...
+            table_html_url = table_html_url[:table_html_url.rfind('/')]
+
+            table_html_a_all = table_soup.find_all('a')
+            for cur_table_html_a in table_html_a_all:
+                if cur_table_html_a.string and 'excel' in cur_table_html_a.string.lower(
+                ):
+                    table_xls_url = cur_table_html_a.get('href')
+                    table_xls_url = _get_abs_url(table_html_url, table_xls_url)
+
+                    print(table_xls_url)
```
Comment (ajaits) [inline scripts/fbi/hate_crime/download_publication_data.py:275]: Pls add more context.

File: scripts/fbi/hate_crime/download_publication_data.py [hunk 14]
```diff
@@ -0,0 +1,363 @@
...
+            _STATUS_DICT['download_urls'][cur_year] = []
+
+            access_url = _find_access_tables_url()
+
+            if access_url:
+                print(access_url)
+                _find_table_html_url()
+                _find_table_xls_url()
+            else:
+                print('No access tables url')
```
Comment (ajaits) [inline scripts/fbi/hate_crime/download_publication_data.py:344]: could add context, `ERROR:` is this is an error as well as year, base url, etc.

File: scripts/fbi/hate_crime/download_publication_data.py [hunk 15]
```diff
@@ -0,0 +1,363 @@
...
+
+def _get_abs_url(base_url: str, cur_url: str) -> str:
+    """Convert relative URL to absolute URL.
+  """
+    ret_url = cur_url
+    if not cur_url.startswith('http'):
+        if cur_url.startswith('/'):
+            ret_url = requests.compat.urljoin(base_url, cur_url)
+        else:
+            ret_url = base_url + '/' + cur_url
```
[the remark this answers is not in the document]
Comment (rbhoot) [reply scripts/fbi/hate_crime/download_publication_data.py:79]: It claims to handle both the cases but on executing the code actually, the expected output was not present and hence the separate handling.

File: scripts/fbi/hate_crime/download_publication_data.py [hunk 16]
```diff
@@ -0,0 +1,363 @@
...
+"""
+The file names of excel files changes across the years.
+The URLs to access table appear in sequence of table number,
+hence code uses enumerate to assign the table number.
+years_with_no_table0 list help make this correction.
+"""
+_YEARS_WITH_NO_TABLE0 = ('2004', '2005', '2006', '2007')
+
+# dictionary to access status across method calls
+# TODO(issue id here) remove this as global variable
```
[the remark this answers is not in the document]
Comment (rbhoot) [reply scripts/fbi/hate_crime/download_publication_data.py:55]: fixed.

File: scripts/fbi/hate_crime/download_publication_data.py (hunk 3 above)
Comment (rbhoot) [reply scripts/fbi/hate_crime/download_publication_data.py:50]: fixed.

File: scripts/fbi/hate_crime/download_publication_data.py [hunk 17]
```diff
@@ -0,0 +1,363 @@
...
+
+flags.DEFINE_integer('start_year',
+                     2005,
+                     'Start year(Min. 2005)',
+                     lower_bound=2005)
+flags.DEFINE_integer('end_year', 2019, 'End year(Min. 2005)', lower_bound=2005)
+flags.DEFINE_boolean('force_fetch', True,
+                     'forces api query and not return cached result')
+flags.DEFINE_string('store_path', './tmp', 'Path to store the output')
+flags.DEFINE_string('cache_path', './tmp/.cache_html/',
```
[the remark this answers is not in the document]
Comment (rbhoot) [reply scripts/fbi/hate_crime/download_publication_data.py:33]: fixed.

File: scripts/fbi/hate_crime/download_publication_data.py [hunk 18]
```diff
@@ -0,0 +1,363 @@
...
+FLAGS = flags.FLAGS
+
+flags.DEFINE_integer('start_year',
+                     2005,
+                     'Start year(Min. 2005)',
+                     lower_bound=2005)
+flags.DEFINE_integer('end_year', 2019, 'End year(Min. 2005)', lower_bound=2005)
+flags.DEFINE_boolean('force_fetch', True,
+                     'forces api query and not return cached result')
+flags.DEFINE_string('store_path', './tmp', 'Path to store the output')
```
[the remark this answers is not in the document]
Comment (rbhoot) [reply scripts/fbi/hate_crime/download_publication_data.py:32]: fixed.

File: scripts/fbi/hate_crime/download_publication_data.py (hunk 1 above)
Comment (rbhoot) [reply scripts/fbi/hate_crime/download_publication_data.py:17]: fixed.

File: scripts/fbi/hate_crime/download_publication_data.py (hunk 4 above)
Comment (rbhoot) [reply scripts/fbi/hate_crime/download_publication_data.py:101]: fixed.

File: scripts/fbi/hate_crime/download_publication_data.py (hunk 5 above)
Comment (rbhoot) [reply scripts/fbi/hate_crime/download_publication_data.py:118]: fixed.

File: scripts/fbi/hate_crime/download_publication_data.py (hunk 6 above)
Comment (rbhoot) [reply scripts/fbi/hate_crime/download_publication_data.py:134]: fixed.

File: scripts/fbi/hate_crime/download_publication_data.py (hunk 7 above)
Comment (rbhoot) [reply scripts/fbi/hate_crime/download_publication_data.py:153]: fixed.

File: scripts/fbi/hate_crime/download_publication_data.py (hunk 10 above)
Comment (rbhoot) [reply scripts/fbi/hate_crime/download_publication_data.py:228]: fixed. thanks for pointing it out.

File: scripts/fbi/hate_crime/download_publication_data.py (hunk 12 above)
Comment (rbhoot) [reply scripts/fbi/hate_crime/download_publication_data.py:243]: fixed.

File: scripts/fbi/hate_crime/download_publication_data.py (hunk 13 above)
Comment (rbhoot) [reply scripts/fbi/hate_crime/download_publication_data.py:275]: fixed.

File: scripts/fbi/hate_crime/download_publication_data.py [hunk 19]
```diff
@@ -0,0 +1,363 @@
...
+    global _STATUS_DICT
+
+    session = requests.Session()
+    session.headers.update(_HEADERS)
+
+    _STATUS_DICT['session'] = session
+
+    _STATUS_DICT['output_dir'] = os.path.expanduser(_STATUS_DICT['output_dir'])
+    if not os.path.exists(_STATUS_DICT['output_dir']):
+        os.makedirs(_STATUS_DICT['output_dir'], exist_ok=True)
```
[the remark this answers is not in the document]
Comment (rbhoot) [reply scripts/fbi/hate_crime/download_publication_data.py:307]: Thanks for pointing out. Fixed.

File: scripts/fbi/hate_crime/download_publication_data.py (hunk 14 above)
Comment (rbhoot) [reply scripts/fbi/hate_crime/download_publication_data.py:344]: fixed.

File: scripts/fbi/hate_crime/download_publication_data.py (hunk 8 above)
Comment (rbhoot) [reply scripts/fbi/hate_crime/download_publication_data.py:166]: fixed.

File: scripts/fbi/hate_crime/download_publication_data.py [hunk 20]
```diff
@@ -0,0 +1,363 @@
...
+                            os.path.join(_STATUS_DICT['output_dir'],
+                                         'url_list.json'), 'w') as fp:
+                        json.dump(_STATUS_DICT['download_urls'], fp, indent=2)
+                    # fetch and store urls
+                    r = requests.get(table_xls_url)
+                    output_path = os.path.join(_STATUS_DICT['output_dir'],
+                                               f"{_STATUS_DICT['cur_year']}")
+                    with open(os.path.join(output_path, f'table_{table_i}.xls'),
+                              'wb') as fp:
+                        fp.write(r.content)
```
[the remark this answers is not in the document]
Comment (rbhoot) [reply scripts/fbi/hate_crime/download_publication_data.py:289]: fixed.

**Comment by rbhoot** [comment]
Thanks for the review @ajaits I have tried to address most the comments. I have avoided using breaks and continue for easier debugging and I usually try to avoid using them. If you want i can make the requested changes.

File: scripts/fbi/hate_crime/download_publication_data.py [hunk 21]
```diff
@@ -0,0 +1,363 @@
...
+        with open(os.path.join(_STATUS_DICT['cache_dir'], 'base.html'),
+                  'r') as fp:
+            base_html = fp.read()
+            base_soup = BeautifulSoup(base_html, features="html.parser")
+
+    for cur_year in _STATUS_DICT['year_list']:
+        _STATUS_DICT['cur_year'] = cur_year
+        year_url = _find_year_url(base_soup, cur_year)
+        if year_url:
+            print(year_url)
```
[the remark this answers is not in the document]
Comment (rbhoot) [reply scripts/fbi/hate_crime/download_publication_data.py:329]: I didn't understand what part is extra?

File: scripts/fbi/hate_crime/download_publication_data.py (hunk 2 above)
Comment (rbhoot) [reply scripts/fbi/hate_crime/download_publication_data.py:43]: fixed. Created issue https://github.com/datacommonsorg/data/issues/639

File: scripts/fbi/hate_crime/download_publication_data.py (hunk 21 above)
[the remark this answers is not in the document]
Comment (ajaits) [reply scripts/fbi/hate_crime/download_publication_data.py]: the print seems to be for debugging and could be removed or would need more text context for a user to understand.

File: scripts/fbi/hate_crime/.gitignore [hunk 22]
```diff
@@ -0,0 +1 @@
+tmp/
```
Comment (ajaits) [inline scripts/fbi/hate_crime/.gitignore:1]: psl add a newline at the end

File: scripts/fbi/hate_crime/README.md [hunk 23]
```diff
@@ -38,6 +38,37 @@ To download data for subset of years
...
+### Notes
+
+- Currently the script is not able to download data for 2004 and Table 13, 14 for 2005
+- By default the extension of saved file is `.xls`. This might cause a problem is extenstions are changed in future.
+- The script tries to find a link to `Access Tables` at one stage. The first instance of it is used if multiple links are found.
+- It is a good idea to check the size or content type(pandas load) of data for a sanity check. If some error was encountered, HTML data might be stored in the file making it unusable with xls applications.
+
+## Examples
+
+To download data from 2005 to 2019
```
Comment (ajaits) [inline scripts/fbi/hate_crime/README.md:62]: pls add a note on where the output files would be store with the commands below.

**Review by ajaits — APPROVED** [review]

**Pull request merged**